docs/source/coding_conventions.rst

```diff
@@ -0,0 +1,119 @@
+Coding Conventions
+==================
+
+In order to increase readability across Tensor Comprehensions written by
+multiple authors and to reduce the amount of surprising behavior, the
+following conventions should be adopted when writing TC. Generally in TC one
+should increment nesting by 4 whitespaces at each level and align tensor names
+and indices where appropriate to make memory access patterns emerge. Since
+these two goals can easily be conflicting, use your best judgement to tradeoff
+between the two goals. Such examples are provided below.
+
+Use indices named after parameters
+----------------------------------
+
+Use upper-case names for parameters and input/output tensors.
+Use lower-case names for indices to match the name of the parameter
+corresponding to the dimension upon which they iterate.
+In other words, prefer:
+
+.. code::
+
+    def copy2d(float(M, N) I) -> (O) {
+        O(m, n) = I(m, n)
+    }
+
+to:
+
+.. code::
+
+    def copy2d(float(M, N) I) -> (O) {
+        O(i, j) = I(i, j)
+    }
+
+Prefix reduction index names with :code:`r_`
+--------------------------------------------
+
+By definition, reduction indices are the ones that appear on the RHS of a TC
+expression but not on the LHS. On larger expressions it can get challenging to easily
+detect the reduction variables by mentally parsing the set of indices on the
+RHS and subtracting the set of indices on the LHS from it. To alleviate such
+issues, name the reduction variables with a :code:`r_` prefix.
+In other words, prefer:
+
+.. code::
+
+    def matmul(float(M, K) A, float(K, N) B) -> (C) {
+        C(m, n) +=! A(m, r_k) * B(r_k, n)
+    }
+
+to:
+
+.. code::
+
+    def matmul(float(M, K) A, float(K, N) B) -> (C) {
+        C(m, n) +=! A(m, k) * B(k, n)
+    }
+
+Filter non-rectangular regions with deta-dependencies
+-----------------------------------------------------
+
+TC semantics only support (hyper-)rectangular iteration spaces. This is a hard
+requirement to make range inference non-ambiguous. To simulate non-rectangular
+iteration spaces, one can use the following:
+
+.. code::
+
+    def matmul(float(M, K) L, float(K, M) U) -> (LU) {
+        LU(m1, m2) +=! (r_k >= m1 and r_k =< m2) ? L(m1, r_k) * U(r_k, m2) : 0
+    }
+
+However, the following is incompatible with range inference and will fail
+the semantic checks in the TC compiler:
+
+.. code::
+
+    def matmul(float(M, K) L, float(K, M) U) -> (LU) {
+        LU(m1, m2) +=! L(m1, r_k) * U(r_k, m2) where r_k in m1:M, r_k in 0:m2+1
+    }
+
+The reader may remark that this is an inefficient way of performing
+matrix-multiplication of triangular matrices.
+Lowering such operations efficient from TC is the subject of future work.
+
+Prefix gradient tensors names with :code:`g_`
+---------------------------------------------
```

Comment:
I can of course make AD follow this pattern, however, provided we disallow mutating inputs, this is not the whole story. Consider a variant of matmul that also has a second output which is the doubled result:
```
def mm(double(M, K) A, double(K, N) B) -> (O, W) {
  O(i, j) +=! A(i, k) * B(k, j)
  W(i, j) = O(i, j) * 2
}
```

So far so good, however if you want to apply reverse mode AD to it, you will need to add gradients w.r.t. W to the gradient w.r.t. O:
```
def grad_mm(double(M, K) A, double(K, N) B,
            double(M, N) O, double(M, N) W,
            double(M, N) seed_d_O, double(M, N) seed_d_W) -> (d_A, d_B) {
  d_O(i, j) = seed_d_d_O(i, j)
  d_O(i, j) += (seed_d_W(i, j) * 2)
  d_A(i, k) = 0
  d_A(i, k) += (d_O(i, j) * B(k, j))
  d_B(k, j) = 0
  d_B(k, j) += (d_O(i, j) * A(i, k))
}
```

Here, you can see that the arguments are gradient "seeds" (this is the name used sometimes in the literature IIRC), and might not reflect the real gradients of a particular value alone, as can be seen in the case of O. On the other hand, is true that the seed is the real gradient of W, which is why my code skips instantiating `d_W`, and uses the seed alone.

If you're writing down the conventions for gradient TCs then it would be good to take that into account. Otherwise, if you're fine with what I did for AD, I can just call them `seed_g_<name>`.

Reply:
I'd prefer `grad_` over `g_`.  Otherwise it smells like the Hungarian notation, and that is a questionable choice.

Reply:
How about @apaszke's ```d_```? It is the actual mathematical notation so that should settle it.

Reply:
@apaszke the seed stuff lgtm, no need to change IMO